sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/OffsetRangeTracker.java

```diff
@@ -262,11 +290,12 @@ public synchronized String toString() {
    */
   @VisibleForTesting
   OffsetRangeTracker copy() {
-    OffsetRangeTracker res = new OffsetRangeTracker(startOffset, stopOffset);
-    res.offsetOfLastSplitPoint = this.offsetOfLastSplitPoint;
-    res.lastRecordStart = this.lastRecordStart;
-    res.done = this.done;
-    res.splitPointsSeen = this.splitPointsSeen;
+    OffsetRangeTracker res =
```

Comment:
Making each get/set method synchronized completely defeats the purpose of synchronization - the purpose of synchronization is usually to make certain changes to multiple variables appear atomic to other threads (sure, there is also the goal of avoiding low-level memory model issues on a single variable, but if that was all that was needed, Java would have the "synchronized" keyword on fields, not on methods). Six synchronized calls with no synchronization between them is not atomic - other threads can intervene while this method is running, and it can return a "copy" that has the six variables copied from six different versions of the current tracker.

Reply:
Got it. Thanks

Reply:
Thanks again for the feedback Eugene. Instead of creating synchronized getters and setters, I created a synchronized lock within the copy method:

    final Object lock = new Object();
    OffsetRangeTracker res = null;
    synchronized(lock) {
      res = new OffsetRangeTracker(this.startOffset, this.stopOffset);
      res.offsetOfLastSplitPoint = this.offsetOfLastSplitPoint;
      res.lastRecordStart = this.lastRecordStart;
      res.done = this.done;
      res.splitPointsSeen = this.splitPointsSeen;
    }
    return res;

This has passed the test cases. Please advise. Thanks.

Reply:
I don't see the changed code. Did you push?

In general the code you quoted in your comment is incorrect. It synchronizes on a local variable, which means it doesn't synchronize at all, because each thread will synchronize on its own variable.

I think to fix this bug you should simply mark the copy() method as "synchronized" and revert all other changes. Does that not work?

Reply:
Hey Eugene, I wanted to verify before pushing changes again. Yes I tried adding synchronized to the copy method, but that did not seem to work. I figured it was some inconsistent synchronization going on with the way that the copied OffsetRangeTracker object's variables were being assigned from the current OffsetRangeTracker.

Reply:
Oh I suppose the problem is in setting the fields on the other tracker - then you need to synchronize on both the current tracker and the other one. Suppose you could do synchronized(this) { synchronized(res) { ... } }.

Reply:
That works! Thanks Eugene.